config/foreman.hiera/family/RedHat-8.yaml

```diff
@@ -8,3 +8,6 @@ apache::mod::ssl::ssl_protocol:
 
 postgresql::globals::manage_dnf_module: true
 postgresql::globals::version: "12"
+
+redis::dnf_module_stream: "6"
+redis::package_ensure: "latest"
```

Comment:
This solves the update problem, but not sure if it's what we want? The dnf module does get reset properly, the package just needs to then get updated as its by default set to 'installed'. It feels right that the puppet layer handles all of this nicely for it. I suppose the other option is to do a hook that calls basically the same puppet code?

hooks/pre/40-update-redis.rb
```
apply_puppet_code("package { 'redis module': ensure => '6', name => 'redis', provider => 'dnfmodule', enable_only => true }")
apply_puppet_code("package { 'redis': ensure => 'latest' })
```

Reply:
This would mean that it *always* ensures latest, not only during the upgrade, which doesn't feel like what we want, indeed.

Reply:
The same would happen in the hook as well, unless we do a check of the Redis version and check for it. Or we have to add this enable/reset code to foreman-maintain for the module to happen during the upgrade process.

Reply:
In the past (el7 postgresql upgrade), we checked if the migration is needed and only executed it once, but that was also a more complicated migration.

Maintain feels like the wrong place - some people omit it and would end in a weird place (and not fully supported upstream anyways).

I'm torn, the current code is so nice and short… and in *most* cases the latest ensure would be a noop anyways.

Reply:
I've debated whether we would want to ensure it for every package (foreman, plugins, etc) in our installer. Still also a bit torn. I mean, ideally we'd have something like `redis::package_ensure: ">= 6.0"` but I don't think that's supported by Puppet.

Reply:
That worked on RPM:

```
redis::package_ensure: ">6"
```

According to Puppet -- https://puppet.com/docs/puppet/7/types/package.html
```
On supported providers, version ranges can also be ensured. For example, inequalities: <2.0.0, or intersections: >1.0.0 <2.0.0.
```

Question is if this works for Debian as well

Reply:
It's in the RH-8 file so this only applies to RPM so I'd say it's not a concern.

Reply:
Interestingly, https://puppet.com/docs/puppet/7/types/package.html#package-provider-features doesn't list the dnf provider as supporting ranges. But if it works…

Reply:
Heh, it does say that, but it works:

```
2022-10-28 18:27:44 [DEBUG ] [configure] Executing: '/bin/dnf list redis --showduplicates | sed -e '1,/Available Packages/ d' | awk '{print $2}''
2022-10-28 18:27:46 [DEBUG ] [configure] Package[redis](provider=dnf): Detected Arch argument in package! - Moving arch to end of version string
2022-10-28 18:27:46 [DEBUG ] [configure] Executing: '/bin/rpm -q redis --nosignature --nodigest --qf %{NAME} %|EPOCH?{%{EPOCH}}:{0}| %{VERSION} %{RELEASE} %{ARCH}\n'
2022-10-28 18:27:46 [DEBUG ] [configure] Package[redis](provider=dnf): Upgrading package redis from version 5.0.3-5.module_el8.4.0+955+7126e393 to 6.2.7-1.module_el8.7.0+1144+ab2666e8
2022-10-28 18:27:46 [DEBUG ] [configure] Facter: resolving fact with user_query: operatingsystemmajrelease
2022-10-28 18:27:46 [DEBUG ] [configure] Facter: Searching fact: operatingsystemmajrelease in file: operatingsystemmajrelease.rb
2022-10-28 18:27:46 [DEBUG ] [configure] Facter: Searching fact: operatingsystemmajrelease in core facts and external facts
2022-10-28 18:27:46 [DEBUG ] [configure] Facter: Loading all internal facts
2022-10-28 18:27:46 [DEBUG ] [configure] Facter: List of resolvable facts: [#<Facter::SearchedFact:0x00000000054042a0 @name="operatingsystemmajrelease", @fact_class=Facts::Rhel::Os::Release, @user_query="operatingsystemmajrelease", @type=:legacy, @file=nil>]
2022-10-28 18:27:46 [DEBUG ] [configure] Facter: Loading external facts
2022-10-28 18:27:46 [DEBUG ] [configure] Facter: fact "operatingsystemmajrelease" has resolved to: 8
2022-10-28 18:27:46 [DEBUG ] [configure] Executing: '/bin/dnf -d 0 -e 1 -y upgrade redis-6.2.7-1.module_el8.7.0+1144+ab2666e8'
2022-10-28 18:27:50 [DEBUG ] [configure] Executing: '/bin/rpm -q redis --nosignature --nodigest --qf %{NAME} %|EPOCH?{%{EPOCH}}:{0}| %{VERSION} %{RELEASE} %{ARCH}\n'
2022-10-28 18:27:50 [INFO  ] [configure] /Stage[main]/Redis::Install/Package[redis]/ensure: ensure changed '5.0.3-5.module_el8.4.0+955+7126e393' to '>=6.0'
2022-10-28 18:27:50 [DEBUG ] [configure] /Package[redis]: The container Class[Redis::Install] will propagate my refresh event
```

Reply:
Ah, interesting, it was intentionally implemented in https://tickets.puppetlabs.com/browse/PUP-10298, but seems never has been documented.

This is present in 6.15.0+, which we depend on anyways. I like!

Reply:
https://github.com/puppetlabs/puppet/pull/8943